### ISSUE TITLE
ESIL: ldur is not working

### ISSUE THREAD

**Issue by YugoCode** [radare/radare2#9197]
_str_ and _ldr_ are working:
```
str x2, [sp + local_18h] 
ldr x0, [sp + local_18h]   
```

But in this combination only _stur_ works:
```
stur x0, [x29 - local_8h] 
ldur x1, [x29 - local_8h]         
```

_stur_ stores the value correctly, but _ldur_ won't touch x1, so x1 keeps it's value. I also checked that x0 != x1.

I noticed this in a couple of iOS binaries. Here is [one](https://github.com/samuelclay/NewsBlur/blob/55c36a24f38c25561e64b4365767547e3945ac1b/clients/ios/NewsBlur.ipa) from an open source project. The function is _"method.FirstTimeUserAddFriendsViewController.setFacebookButton:"_.

_______

> radare2 2.3.0-git 17047 @ darwin-x86-64 git.2.2.0-236-gbec01d21a
> commit: bec01d21ac735ca1b25842a27e3c2389d310871a build: 2018-01-15__01:01:25

**Comment by radare** [comment]
Mind to send a fix? Just edit the esil expression

**Comment by YugoCode** [comment]
I looked into _libr/anal/p/anal_arm_cs.c_ for it. 

There are two functions, which handle the load instructions: _anop64()_ and _analop64_esil()_.
Both handle the load instructions generic, so if _ldr_ works, then _ldur_ should also work. 
```
	case ARM64_INS_LDUR:
	case ARM64_INS_LDR:
	case ARM64_INS_LDRSB:
	case ARM64_INS_LDRB:
	case ARM64_INS_LDRSW:
             do_stuff...
```

I think it depends on the used registers, because all the _ldur_ instructions I've seen in my binaries use _x29_ and the _ldr_ instructions not. 

Then maybe it's this part?
```
if (REGBASE64(1) == ARM64_REG_X29) {
			op->stackop = R_ANAL_STACK_GET;
			op->stackptr = MEMDISP64(1);
		} else {
                            ...
```

I'm really not sure, so unfortunately I cannot fix it 😿

**Comment by radare** [comment]
But... r u taklking about emulation? Because then u have to check the esil expression. Press O in visual mode to toggle that. Or just use the so command.

Can you provide a testcase to rerproduce this?

**Comment by radare** [comment]
By a testcase i mean an r2 oneliner to run the issue and check the output

**Comment by giuscri** [comment]
Any hint on how to fix this?

**Comment by giuscri** [comment]
I think I can handle this.

Apparently, the ldru instruction is _not_ assembled.

```
$ rasm2 -a arm -b 64 'ldur x1, [x29, 8]'
Cannot assemble 'ldur x1, [x29, 8]' at line 3
invalid
```

Adding needed code at `libr/asm/arch/arm/armass64.c:arm64ass` should suffice.

I'm diving into that in the next hours.

**Comment by Maijin** [comment]
arm.as (*arm* handmade assembler) is one of the 2 arm assemblers available in radare2 and is updated frequently.  **Please add a test case so we can fix arm.as in https://github.com/radare/radare2-regressions/blob/master/t.asm/arm/capstone/arm_asm** You can do so with github web editor without even cloning the repo.

You can contribute to arm.as by completing the following files: 
* https://github.com/radare/radare2/blob/master/libr/asm/arch/arm/armass.c
* https://github.com/radare/radare2/blob/master/libr/asm/arch/arm/armass64.c

You can also use keystone(http://keystone-engine.org) within radare2/rasm2 by installing the radare2 plugin via r2pm : 

```
	r2pm -i keystone-lib
	r2pm -i keystone
then
	rasm2 -a arm.ks…
or in radare2 session
	e asm.assembler = arm.ks
```

**Comment by YugoCode** [comment]
I still have the same problem I described in my first post.

After setting up the vm (with _aei, aeim, aeip_) and stepping through the instructions _ldur_ will not load the value into the register. 

I tried the _rasm2_ oneliner from @giuscri and it gets assembled, but this does not help me when emulating code.

So should we open the issue again?

```
radare2 2.4.0-git 17316 @ darwin-x86-64 git.2.2.0-502-g9c12c4fe2
commit: 9c12c4fe2021fb035797e474c94a9a05c62522c7 build: 2018-02-19__18:32:40
```

**Comment by giuscri** [comment]
Can you share another binary? I don't know how to work with `.ipa` files - to my knowledge I need an iOS device (?).

Only by writing and executing the instructions, I can't reproduce:
```
$ r2 -b64 -aarm -e cfg.fortunes=false -
[0x00000000]> "wa stur x0, [x29, 8]; ldur x1, [x29, 8]"
Written 8 byte(s) (stur x0, [x29, 8]; ldur x1, [x29, 8]) = wx a08300f8a18340f8
[0x00000000]> aei
[0x00000000]> aeim
[0x00000000]> aeip
[0x00000000]> aer x0=0x1122334455667788
[0x00000000]> aes
[0x00000004]> aes
[0x00000008]> aer x1
0x1122334455667788
```

**Comment by Maijin** [comment]
r2 ipa://file.ipa doesn't work?

**Comment by YugoCode** [comment]
Sure, a _.ipa_ file is just a _.zip_ archive. So you can just unzip it, but might need to rename the _.ipa_ to _.zip_ before. 

When you extracted the archive, you should find a Payload folder. Inside this folder is a file with a _.app_ file extension. This is a so called [bundle](https://en.wikipedia.org/wiki/Bundle_(macOS)). The binary is located inside this bundle. Under MacOS you can open it by right clicking and selecting "Show Package Content". On other operating systems it should just be a folder. 

I uploaded the binary from the .ipa I mentioned in my first post, if you had any problems finding it: 
[NewsBlur_binary.zip](https://github.com/radare/radare2/files/1738121/NewsBlur_binary.zip)

**Comment by YugoCode** [comment]
@Maijin 

> r2 ipa://NewsBlur.ipa

This gives me:
```
ZIP File Error: File could not be opened file name.
Warning: File did not exist, creating a new one.
```

And the file could not be loaded, so there are no classes, imports etc.

**Comment by Maijin** [comment]
This will need another update CC @radare

**Comment by YugoCode** [comment]
@Maijin Sorry! It works! The .ipa I tried was corrupted or something.

So @giuscri `r2 ipa://NewsBlur.ipa` is the easier way.

**Comment by YugoCode** [comment]
@giuscri This is a fat binary, so there is a 32-Bit ARMv7 and a 64-Bit ARMv8 binary. You can find the _ldur_ instruction only in the 64-Bit binary. I don't know how to load the 64-Bit binary with a command, but you can use `rabin2 -x fatBinary` in order to extract the single binaries.

For the 64-Bit binary of NewsBlur you can find the _ldur_ instruction in _"method.FirstTimeUserAddFriendsViewController.setFacebookButton:"_. Set a value in _x0_ before the _"stur x0, [x29 - local_8h]"_ and then for me the  instruction _"ldur x1, [x29 - local_8h]"_ won't load the value into _x1_.

**Comment by giuscri** [comment]
OK, binary should be [NewsBlur.arm_64.1.zip](https://github.com/radare/radare2/files/1738253/NewsBlur.arm_64.1.zip)

Analysis is taking really too long on my machine. Where should I seek for `method.FirstTimeUserAddFriendsViewController.setFacebookButton:`?

**Comment by giuscri** [comment]
btw, I found a gadget by which I can reproduce the bug
```
[0x1000b2acc]> pd
            0x1000b2acc      a0831cf8       stur x0, [x29, -0x38]
            0x1000b2ad0      a1835cf8       ldur x1, [x29, -0x38]
            ...
```
It seems like the problem is a wrong ESIL expression: ldur get translated into `x1,x29,56,-,=[8]`. Actually the address computed for the load is `0x38-x29`, _not_ `x29-0x38`.

You can actually reproduce the bug via
```
$ rasm2 -b64 -aarm "ldur x1, [x29, -0x38]"|rasm2 -b64 -aarm -d -E -
x1,x29,56,-,=[8]
```

**Comment by YugoCode** [comment]
> Analysis is really taking too long on my machine. Where should I seek for method.FirstTimeUserAddFriendsViewController.setFacebookButton:?

I don't run an analysis. I open the binary and then do:
```
s method.FirstTimeUserAddFriendsViewController.setFacebookButton:
aei 
aeim
aeip
V!
```
Now you're in visual mode and can press the 's' key to step forward and use the ':' key to run commands like setting _x0_ with `aer`. _Press the enter key to cancel the command view and return to the visual view._ If you want to use `pdf` in order to view the disassembly you can analyze just this one function with `af`. Of course you need to seek to the function before you analyze it. But you don't need to analyze the whole binary.

> It seems like the problem is a wrong ESIL expression

Do you know how to fix it?

______
And thank you for investigating in this ❤️

**Comment by giuscri** [comment]
Tried to fix to that in a #9442. Apparently the problem was there also for `ldr` instruction. Negative offset is causing troubles.
